pkg/kor/clusterroles.go

```diff
@@ -0,0 +1,133 @@
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+)
+
+func retrieveUsedClusterRoles(clientset kubernetes.Interface, namespace string, filterOpts *filters.Options) ([]string, error) {
+	// Get a list of all role bindings in the specified namespace
+	roleBindings, err := clientset.RbacV1().RoleBindings(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list role bindings in namespace %s: %v", namespace, err)
+	}
+
+	usedClusterRoles := make(map[string]bool)
+
+	for _, rb := range roleBindings.Items {
+		if pass, _ := filter.Run(filterOpts); pass {
+			continue
+		}
+		usedClusterRoles[rb.RoleRef.Name] = true
+		if rb.RoleRef.Kind == "ClusterRole" {
+			usedClusterRoles[rb.RoleRef.Name] = true
+		}
+	}
+
+	// Get a list of all cluster role bindings in the specified namespace
+	clusterRoleBindings, err := clientset.RbacV1().ClusterRoleBindings().List(context.TODO(), metav1.ListOptions{})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to list cluster role bindings in namespace %s: %v", namespace, err)
+	}
+
+	for _, crb := range clusterRoleBindings.Items {
+		if pass, _ := filter.Run(filterOpts); pass {
+			continue
+		}
+		usedClusterRoles[crb.RoleRef.Name] = true
+
+		usedClusterRoles[crb.RoleRef.Name] = true
+	}
+
+	var usedClusterRoleNames []string
+	for role := range usedClusterRoles {
+		usedClusterRoleNames = append(usedClusterRoleNames, role)
+	}
+
+	return usedClusterRoleNames, nil
+}
+
+func retrieveClusterRoleNames(clientset kubernetes.Interface) ([]string, error) {
+	clusterRoles, err := clientset.RbacV1().ClusterRoles().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	names := make([]string, 0, len(clusterRoles.Items))
+	for _, clusterRole := range clusterRoles.Items {
+		if clusterRole.Labels["kor/used"] == "true" {
+			continue
+		}
+
+		names = append(names, clusterRole.Name)
+	}
+	return names, nil
+}
+
+// func processNamespaceRoles(clientset kubernetes.Interface, namespace string, filterOpts *FilterOptions) ([]string, error) {
+func processNamespaceClusterRoles(clientset kubernetes.Interface, namespace string, filterOpts *filters.Options) ([]string, error) {
+	usedClusterRoles, err := retrieveUsedClusterRoles(clientset, namespace, filterOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	usedClusterRoles = RemoveDuplicatesAndSort(usedClusterRoles)
+
+	clusterRoleNames, err := retrieveClusterRoleNames(clientset)
+	if err != nil {
+		return nil, err
+	}
+
+	diff := CalculateResourceDifference(usedClusterRoles, clusterRoleNames)
+	return diff, nil
+
+}
+
+func GetUnusedClusterRoles(filterOpts *filters.Options, clientset kubernetes.Interface, outputFormat string, opts Opts) (string, error) {
+	var outputBuffer bytes.Buffer
+	namespaces := filterOpts.Namespaces(clientset)
+	response := make(map[string]map[string][]string)
+
+	for _, namespace := range namespaces {
```

Comment:
Cluster roles are not namespaced. The behavior should be to look at all namespaces for roleBindings and clusterRoleBinding wether they contain the clusterRole

Reply:
+1

Reply:
@yonahd 
That was quite a refactor, fixed :)

Reply:
Thanks!